Misleading message when attempting to call methods on a raw pointer

When trying to call a method on a raw pointer Rust compiler says there's no such method.

It would be better to say that there is such method, but the pointer has to be explicitly dereferenced first.

``` rust
let x = 8u8;
let y = &x;
let z: *const u8 = &x;
println!("{} {}", 
    y.to_string(), // OK
    z.to_string()  // Not OK
);
```

> error: type `*const u8` does not implement any method in scope named `to_string`

This is _technically correct_, but unexpected and help hint for this error suggests wrong solution. 

I've read this message as "yada yada _no such method_", so I went to check whether I've spelled the method correctly, whether I've put the method in the appropriate `impl` block, etc. — all completely irrelevant to this error.

I'd find something like that more helpful:

> error: method `to_string` requires type `u8`, but `*const u8` cannot be automatically dereferenced. hint: use `as_ref()`

Triage: the error message has improved
```
error[E0599]: no method named `to_string` found for type `*const u8` in the current scope
 --> src/main.rs:8:11
  |
8 |         z.to_string()  // Not OK
  |           ^^^^^^^^^
  |
  = note: the method `to_string` exists but the following trait bounds were not satisfied:
          `*const u8 : std::string::ToString`
```
`rustc: 1.32.0`

--- a/src/librustc_typeck/check/method/suggest.rs
+++ b/src/librustc_typeck/check/method/suggest.rs
@@ -324,6 +324,11 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                                 );
                             }
                         }
+                        if let ty::RawPtr(_) = &actual.sty {
+                            err.note("try using `<*const T>::as_ref()` to get a reference to the \
+                                      type behind the pointer: https://doc.rust-lang.org/std/\
+                                      primitive.pointer.html#method.as_ref");
+                        }
                         err
                     }
                 } else {

--- a/src/test/ui/issues/issue-21596.rs
+++ b/src/test/ui/issues/issue-21596.rs
@@ -1,0 +1,5 @@
+fn main() {
+    let x = 8u8;
+    let z: *const u8 = &x;
+    println!("{}", z.to_string());  //~ ERROR E0599
+}

--- a/src/test/ui/issues/issue-21596.stderr
+++ b/src/test/ui/issues/issue-21596.stderr
@@ -1,0 +1,13 @@
+error[E0599]: no method named `to_string` found for type `*const u8` in the current scope
+  --> $DIR/issue-21596.rs:4:22
+   |
+LL |     println!("{}", z.to_string());
+   |                      ^^^^^^^^^
+   |
+   = note: try using `<*const T>::as_ref()` to get a reference to the type behind the pointer: https://doc.rust-lang.org/std/primitive.pointer.html#method.as_ref
+   = note: the method `to_string` exists but the following trait bounds were not satisfied:
+           `*const u8 : std::string::ToString`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0599`.

